interfaces/builtin/dbus_app.go

```diff
@@ -0,0 +1,244 @@
+			snippets.Write(snippet)
+
+			for _, name := range names {
+				// well-known DBus name-specific permanent slot
+				// policy
+				snippet := getAppArmorIndividualSnippet([]byte(dbusAppPermanentSlotAppArmorIndividual), bus, name)
+				snippets.Write(snippet)
+
+				if release.OnClassic {
+					// classic-only policy
+					snippet := getAppArmorIndividualSnippet([]byte(dbusAppPermanentSlotAppArmorIndividualClassic), bus, name)
+					snippets.Write(snippet)
+				}
+			}
+		}
+		//fmt.Printf("DEBUG - PERMANENT SLOT:\n %s\n", snippets.Bytes())
+		return snippets.Bytes(), nil
+	case interfaces.SecuritySecComp:
+		return []byte(dbusAppPermanentSlotSecComp), nil
+	}
+	return nil, nil
+}
+
+func (iface *DbusAppInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+// Obtain yaml-specified DBus well-known names by bus
+func (iface *DbusAppInterface) getBusNames(attribs map[string]interface{}) (map[string][]string, error) {
+	busNames := make(map[string][]string)
+	for attr := range attribs {
+		bus := attr
+		if bus != "session" && bus != "system" {
+			return nil, fmt.Errorf("bus must be one of 'session' or 'system'")
+		}
+		busNamesList, ok := attribs[bus].([]interface{})
+		if !ok {
+			return nil, fmt.Errorf("bus attribute is not a list")
+		}
+
+		for _, item := range busNamesList {
+			busName, ok := item.(string)
+			if !ok {
+				return nil, fmt.Errorf("session element is not a string")
+			}
+
+			// https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names
+			if !ok || len(busName) == 0 {
+				return nil, fmt.Errorf("bus name must be set")
+			} else if len(busName) > 255 {
+				return nil, fmt.Errorf("bus name is too long (must be <= 255)")
+			}
+
+			validBusName := regexp.MustCompile("^[a-zA-Z][a-zA-Z0-9_-]*(\\.[a-zA-Z0-9_-]+)+$")
+			if !validBusName.MatchString(busName) {
+				return nil, fmt.Errorf("invalid bus name: %q", busName)
+			}
+
+			busNames[bus] = append(busNames[bus], busName)
+		}
+	}
+
+	return busNames, nil
+}
+
+// Determine AppArmor dbus abstraction to use based on bus
+func getAppArmorAbstraction(bus string) (string, error) {
+	var abstraction string
+	if bus == "system" {
+		abstraction = "dbus-strict"
+	} else if bus == "session" {
+		abstraction = "dbus-session-strict"
+	} else {
+		return "", fmt.Errorf("unknown abstraction for specified bus '%s'", bus)
+	}
+	return abstraction, nil
+}
+
+// Calculate individual snippet policy based on bus and name
+func getAppArmorIndividualSnippet(policy []byte, bus string, name string) []byte {
+	old := []byte("###DBUS_BIND_BUS###")
+	new := []byte(bus)
+	snippet := bytes.Replace(policy, old, new, -1)
+
+	old = []byte("###DBUS_BIND_NAME###")
+	new = []byte(name)
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus path
+	dot_re := regexp.MustCompile("\\.")
```

Comment:
This seems very module specific. Ie, trying to convert `org.test-session1` to `/org/test-session1{,/**}`. This for writing a particular glob rule for policy and abstracting it out likely has little benefit and would make policy code reviews more difficult.